File: tests/refs/coin_change.py
"""Reference: minimum coin change DP (LC 322).

Returns the minimum number of coins making `amount`, or -1 if impossible.
"""

from __future__ import annotations
import math
from typing import List


def coinChange(coins: List[int], amount: int) -> int:
    INF = math.inf
    dp: List[float] = [INF] * (amount + 1)
    dp[0] = 0
    for a in range(1, amount + 1):
        for c in coins:
            if c <= a and dp[a - c] + 1 < dp[a]:
                dp[a] = dp[a - c] + 1
    return -1 if dp[amount] == INF else int(dp[amount])

File: tests/refs/CoinChange.java
/*
 * Reference Java implementation for tests/cases/coin_change.json.
 * Minimum coins to make `amount`, or -1 if impossible (LC 322).
 */
import java.util.Arrays;

public class CoinChange {
    public static long coinChange(long[] coins, long amount) {
        int amt = (int) amount;
        int INF = Integer.MAX_VALUE;
        int[] dp = new int[amt + 1];
        Arrays.fill(dp, INF);
        dp[0] = 0;
        for (int a = 1; a <= amt; ++a) {
            for (long c : coins) {
                if (c <= a && dp[a - (int) c] != INF && dp[a - (int) c] + 1 < dp[a]) {
                    dp[a] = dp[a - (int) c] + 1;
                }
            }
        }
        return dp[amt] == INF ? -1 : dp[amt];
    }
}
